### PR TITLE
[SYCLomatic] Enable migration of cuda_compile_ptx() with cmake helper function sycl_compile_device() introduced

### PR DESCRIPTION
**Pull request by tomflinda**
1. Introduce cmake helper file sycl_helper.cmake to fill the gap between cuda cmake synax and sycl part.
2. Implement migration of cmake syntax cuda_compile_ptx() with helper function dpct_compile_sycl_code() introduced.

### CODE REVIEW

**Review by zhiweij1 — APPROVED** [review]

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/test/dpct/cmake_migation/case_001/expected.txt [hunk 1]
```diff
@@ -16,6 +21,16 @@ include_directories(
 add_executable(foo-bar ${SOURCES})
 project(foo )
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsycl")
+find_program(dpct_bin_path NAMES dpct PATHS)
```
Comment (zhimingwang36) [inline clang/test/dpct/cmake_migation/case_001/expected.txt:24]:
```suggestion
find_program(dpct_binnary NAMES dpct PATHS)
```

File: clang/test/dpct/cmake_migation/case_001/expected.txt [hunk 2]
```diff
@@ -16,6 +21,16 @@ include_directories(
...
+get_filename_component(dpct_path ${dpct_bin_path} DIRECTORY)
```
Comment (zhimingwang36) [inline clang/test/dpct/cmake_migation/case_001/expected.txt]:
```suggestion
get_filename_component(bin_path_of_dpct ${dpct_bin_path} DIRECTORY)
```

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml [hunk 3]
```diff
@@ -16,6 +16,10 @@
   Out: |
    project(${arg0} ${argN})
    set(CMAKE_CXX_FLAGS "\${CMAKE_CXX_FLAGS} -fsycl")
+   find_program(dpct_bin_path NAMES dpct PATHS)
+   get_filename_component(dpct_path \${dpct_bin_path} DIRECTORY)
+   set(cmake_helper_file_path "\${dpct_path}/../cmake/sycl_helper.cmake")
```
Comment (zhimingwang36) [inline clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml]:
```suggestion
   set(cmake_helper_file_path "\${dpct_path}/../cmake/dpct.cmake")
```

File: clang/tools/dpct/cmake/sycl_helper.cmake [hunk 4]
```diff
@@ -0,0 +1,69 @@
+macro(GET_SOURCES _sources)
```
Comment (zhimingwang36) [inline clang/tools/dpct/cmake/sycl_helper.cmake]: //==---- dpct.hpp ---------------------------------*- C++ -*----------------==//
//
// Copyright (C) Intel Corporation
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
// See https://llvm.org/LICENSE.txt for license information.
//
//===----------------------------------------------------------------------===//

File: clang/tools/dpct/CMakeLists.txt [hunk 5]
```diff
@@ -42,3 +42,18 @@ if(UNIX)
 endif()
 
 add_subdirectory(DpctOptRules)
+
+set(dpct_cmake_helper_file
```
Comment (zhimingwang36) [inline clang/tools/dpct/CMakeLists.txt:46]:
```suggestion
set(dpct_cmake_file
```

File: clang/tools/dpct/cmake/sycl_helper.cmake (hunk 4 above)
Comment (zhimingwang36) [inline clang/tools/dpct/cmake/sycl_helper.cmake]:
```suggestion
macro(DPCT_GET_SOURCES _sources)
```

File: clang/tools/dpct/cmake/sycl_helper.cmake [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  set( ${_sources} )
+  foreach(arg ${ARGN})
+    # Assume arg is a source file
+    list(APPEND ${_sources} ${arg})
+  endforeach()
+endmacro()
+
+macro(INNER_SRS_WRAPPER sycl_target generated_files)
```
Comment (zhimingwang36) [inline clang/tools/dpct/cmake/sycl_helper.cmake:9]:
```suggestion
macro(INNER_SOURCE_WRAPPER sycl_target generated_files)
```

File: clang/tools/dpct/cmake/sycl_helper.cmake [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+endmacro()
+
+macro(SYCL_COMPILE_BASE sycl_target generated_files)
+
+  set(_sycl_target "${sycl_target}")
+  GET_SOURCES(_sources ${ARGN})
+
+  
+  # Create custom command for each cpp source file
+  INNER_SRS_WRAPPER( ${_sycl_target} _generated_files ${_sources})
```
Comment (zhimingwang36) [inline clang/tools/dpct/cmake/sycl_helper.cmake]:
```suggestion
  DPCT_CREATE_BUILD_COMMAND( ${_sycl_target} _generated_files ${_sources})
```

File: clang/tools/dpct/cmake/sycl_helper.cmake [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+  set( ${generated_files} ${_generated_files})
+endmacro()
+
+macro(SYCL_COMPILE_DEVICE generated_files)
```
Comment (zhimingwang36) [inline clang/tools/dpct/cmake/sycl_helper.cmake]:
```suggestion
macro(DPCT_COMPILE_SYCL_CODE generated_files)
```

File: clang/tools/dpct/cmake/sycl_helper.cmake [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        message(FATAL_ERROR "Only support cpp file.")
+    endif()
+
+    math(EXPR _counter "${_counter} + 1")
+  endforeach()
+
+  set(${generated_files} ${_generated_files})
+endmacro()
+
+macro(SYCL_COMPILE_BASE sycl_target generated_files)
```
Comment (zhimingwang36) [inline clang/tools/dpct/cmake/sycl_helper.cmake]:
```suggestion
macro(DPCT_COMPILE_SYCL_CODE_IMP sycl_target generated_files)
```

File: clang/tools/dpct/cmake/sycl_helper.cmake (hunk 8 above)
Comment (zhimingwang36) [inline clang/tools/dpct/cmake/sycl_helper.cmake]: Please add comment for the last macro, as it is expected to be used out size.

File: clang/CMakeLists.txt [hunk 10]
```diff
@@ -916,6 +916,7 @@ if (NOT CLANG_BUILT_STANDALONE)
       install-dpct-binary
       install-dpct-opt-rules
       install-dpct-cmake-rules
+      install-cmake-file
```
Comment (zhimingwang36) [inline clang/CMakeLists.txt]: install-dpct-cmake-file

File: clang/test/dpct/cmake_migation/case_023/case_023_cuda_compile_ptx.cpp [hunk 11]
```diff
@@ -0,0 +1,10 @@
+// RUN: rm -rf %T && mkdir -p %T
+// RUN: cd %T
+// RUN: cp %S/input.cmake ./input.cmake
+// RUN: dpct -in-root ./ -out-root out  ./input.cmake --rule-file=%S/../../../../tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml --migrate-cmake-script-only --cuda-include-path="%cuda-path/include"
```
Comment (zhimingwang36) [inline clang/test/dpct/cmake_migation/case_023/case_023_cuda_compile_ptx.cpp:4]: Please follow up to refine: 
1. if --migrate-cmake-script-only used, then  --cuda-include-path="%cuda-path/include" is not needed. 
2. cmake_script_migration_rule.yaml should be loaded auto

Comment (tomflinda) [reply clang/test/dpct/cmake_migation/case_023/case_023_cuda_compile_ptx.cpp:4]: Will submit a separate PR to refine them

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**